src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageSearch/PackageSearchResultJsonPrinter.cs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using NuGet.CommandLine.XPlat.Commands.PackageSearch;
+using NuGet.Configuration;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+
+namespace NuGet.CommandLine.XPlat
+{
+    internal class PackageSearchResultJsonPrinter : IPackageSearchResultRenderer
+    {
+        private ILoggerWithColor _logger;
+        private const int LineSeparatorLength = 40;
+        private static readonly string SourceSeparator = new('*', LineSeparatorLength);
+        private PackageSearchVerbosity _verbosity;
+        private List<PackageSearchResult> _packageSearchResults;
+
+        public PackageSearchResultJsonPrinter(ILoggerWithColor loggerWithColor, PackageSearchVerbosity verbosity)
+        {
+            _logger = loggerWithColor;
+            _verbosity = verbosity;
+        }
+        public async void Add(PackageSource source, IEnumerable<IPackageSearchMetadata> completedSearch)
+        {
+            PackageSearchResult packageSearchResult = new PackageSearchResult(source.Name);
+
+            foreach (IPackageSearchMetadata metadata in completedSearch)
+            {
+                if (_verbosity == PackageSearchVerbosity.Minimal)
+                {
+                    packageSearchResult.AddPackage(new PackageSearchResultPackageMinimal(metadata));
+                }
+                else if (_verbosity == PackageSearchVerbosity.Detailed)
+                {
+                    PackageDeprecationMetadata packageDeprecationMetadata = await metadata.GetDeprecationMetadataAsync();
+                    packageSearchResult.AddPackage(new PackageSearchResultPackageDetailed(metadata, packageDeprecationMetadata?.Message));
+                }
+                else
+                {
+                    packageSearchResult.AddPackage(new PackageSearchResultPackageNormal(metadata));
+                }
+            }
+
+            _packageSearchResults.Add(packageSearchResult);
+        }
+
+        public void Add(PackageSource source, string error)
+        {
+            _logger.LogMinimal(SourceSeparator);
```

Comment:
Ideally you don't take a logger in the renderer.

Reply:
In the situation we have multiple sources, we want to log the search results of the source that returns first immediately. That will make sure that a slow source among the sources will not slow down the whole search operation. As a result, the renderer takes a logger so that it can log the search result from a source immediately. 

We could have the renderer return a search result string it constructed at the end and do the logging in the `PackageSearchRunner` class. However, that would mean we have to wait for all the sources to return.

Reply:
This is the json printer. 
Do you want to log with the regular logger there?